javascript/ql/src/semmle/javascript/dataflow/internal/InterProceduralTypeInference.qll

```diff
@@ -190,3 +127,119 @@ private class AnalyzedThisInPropertyFunction extends AnalyzedThisExpr {
   }
 }
 
+/**
+ * Holds if the execution of function `f` may complete normally without
+ * encountering a `return` or `throw` statement.
+ *
+ * Note that this is an overapproximation, that is, the predicate may hold
+ * of functions that cannot actually complete normally, since it does not
+ * account for `finally` blocks and does not check reachability.
+ */
+predicate mayReturnImplicitly(Function f) {
+  exists (ConcreteControlFlowNode final |
+    final.getContainer() = f and
+    final.isAFinalNode() and
+    not final instanceof ReturnStmt and
+    not final instanceof ThrowStmt
+  )
+}
+
+/**
+ * A call with inter-procedural type inference for the return value.
+ */
+abstract class CallWithAnalyzedReturnFlow extends DataFlow::CallNode, DataFlow::AnalyzedValueNode {
+
+  /**
+   * Gets a called function.
+   */
+  abstract Function getAFunction();
+  
+  /**
+   * Gets a return value for this call.
+   */
+  AbstractValue getAReturnValue() {
+    exists (Function f | f = getAFunction() |
+      if f.isGenerator() or f.isAsync() then
+        result = TAbstractOtherObject()
+      else (
+        // explicit return value
+        result = f.getAReturnedExpr().analyze().getALocalValue()
+        or
+        // implicit return value
+        (
+          // either because execution of the function may terminate normally
+          mayReturnImplicitly(f)
+          or
+          // or because there is a bare `return;` statement
+          exists (ReturnStmt ret | ret = f.getAReturnStmt() | not exists(ret.getExpr()))
+        ) and
+        result = TAbstractUndefined()
+      )
+    )
+  }
+
+  override AbstractValue getALocalValue() {
+    result = getAReturnValue()
+  }
+}
+
+/**
+ * Flow analysis for the return value of IIFEs.
+ */
+private class IIFEWithAnalyzedReturnFlow extends CallWithAnalyzedReturnFlow {
+  
+  ImmediatelyInvokedFunctionExpr iife;
+  
+  IIFEWithAnalyzedReturnFlow() {
+    this.asExpr() = iife.getInvocation()
+  }
+  
+  override Function getAFunction() {
+    result = iife
+  }
+  
+}
+
+/** A function that only is used locally, making it amenable to type inference. */
+class LocalFunction extends Function {
+
+  DataFlow::Impl::ExplicitInvokeNode invk;
+
+  LocalFunction() {
+    this instanceof FunctionDeclStmt and
+    exists (LocalVariable v, Expr callee |
+      callee = invk.getCalleeNode().asExpr() and
+      v = getVariable() and
+      v.getAnAccess() = callee and
+      forall(VarAccess o | o = v.getAnAccess() | o = callee) and
```

Comment:
Is `count(v.getAnAccess()) = 1` worse in terms of performance?

Reply:
Yes. It was a lot worse in the very first performance evaluation of this PR. It might of course have changed now, do you want a new evaluation?

Reply:
No, just wondering.